tls/s2n_x509_validator.c

```diff
@@ -436,26 +436,59 @@ S2N_RESULT s2n_x509_validator_check_cert_preferences(struct s2n_connection *conn
      *# "signature_algorithms" extension also applies to signatures appearing in
      *# certificates.
      */
-    if (security_policy->certificate_signature_preferences == NULL) {
+    if (security_policy->certificate_signature_preferences == NULL
+            && security_policy->certificate_key_preferences == NULL) {
         return S2N_RESULT_OK;
     }
 
     struct s2n_cert_info info = { 0 };
     RESULT_GUARD(s2n_openssl_x509_get_cert_info(cert, &info));
 
-    if (info.self_signed) {
-        return S2N_RESULT_OK;
-    }
-
     /* Ensure that the certificate signature does not use SHA-1. While this check
      * would ideally apply to all connections, we only enforce it when certificate
      * preferences exist to stay backwards compatible.
      */
-    if (conn->actual_protocol_version == S2N_TLS13) {
+    if (!info.self_signed && conn->actual_protocol_version == S2N_TLS13) {
         RESULT_ENSURE(info.signature_digest_nid != NID_sha1, S2N_ERR_CERT_UNTRUSTED);
     }
 
-    RESULT_ENSURE_OK(s2n_security_policy_validate_cert_signature(security_policy, &info), S2N_ERR_CERT_UNTRUSTED);
+    if (!info.self_signed) {
+        RESULT_ENSURE_OK(s2n_security_policy_validate_cert_signature(security_policy, &info), S2N_ERR_CERT_UNTRUSTED);
+    }
+    RESULT_ENSURE_OK(s2n_security_policy_validate_cert_key(security_policy, &info), S2N_ERR_CERT_UNTRUSTED);
+
```

Comment:
I'm a little worried that your choice to make the usage error the default and override for a protocol error is going to lead to mistakes. It's not a one-off anymore. Maybe these functions should take an error as an argument?

Reply:
We've had this problem before: even if your method shouldn't just be called with RESULT_GUARD, someone is eventually going to just call it with RESULT_GUARD.